muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/guider/GuiderRESTService.java

```diff
@@ -1045,4 +1086,175 @@ else if (value instanceof Double) {
+    String content;
+    if (StringUtils.isEmpty(entity.getMessage())) {
+      content = localeController.getText(sessionController.getLocale(), "rest.workspace.joinWorkspace.joinNotification.counselor.content");
+      content = MessageFormat.format(content, loggedUserName, studentLink, workspaceLink);
+    } else {
+      content = localeController.getText(sessionController.getLocale(), "rest.workspace.joinWorkspace.joinNotification.counselor.contentwmessage");
+      String blockquoteMessage = String.format("<blockquote>%s</blockquote>", entity.getMessage());
+      content = MessageFormat.format(content, loggedUserName, studentLink, workspaceLink, blockquoteMessage);
+    }
+
+    for (MessagingWidget messagingWidget : messagingWidgets) {
+      // TODO: Category?
+      messagingWidget.postMessage(userSchoolDataIdentifier.getUserEntity(), "message", caption, content, recipients);
+    }
+
+    List<String> recipientEmails = new ArrayList<>(recipients.size());
+    for (UserEntity recipient : recipients){
+     String teacherEmail = userEmailEntityController.getUserDefaultEmailAddress(recipient, false);
+     if (StringUtils.isNotBlank(teacherEmail)) {
+       recipientEmails.add(teacherEmail);
+     }
+    }
+    if (!recipientEmails.isEmpty()) {
+      mailer.sendMail(MailType.HTML, recipientEmails, caption, content);
+    }
+    List<StudyActivityItemRestModel> restItems = new ArrayList<StudyActivityItemRestModel>();
+    
+    SearchProvider searchProvider = getProvider("elastic-search");
+    if (searchProvider != null) {
+      SearchResult searchResult =  searchProvider.findWorkspace(workspaceIdentifier);
+      
+      if (searchResult.getTotalHitCount() > 0) {
+        List<Map<String, Object>> results = searchResult.getResults();
+        Map<String, Object> match = results.get(0);
+
+        @SuppressWarnings("unchecked")
+        List<Map<String, Object>> subjectList = (List<Map<String, Object>>) match.get("subjects");
+        for (Map<String, Object> s : subjectList) {
+          StudyActivityItemRestModel item = new StudyActivityItemRestModel();
+
+          item.setCourseName(workspaceName);
+          item.setCourseId(workspaceEntity.getId());
+          item.setCourseNumber((Integer) s.get("courseNumber"));
+          item.setSubject((String) s.get("subjectCode"));
+          item.setStatus(StudyActivityItemStatus.ONGOING);
+          item.setSubjectName((String) s.get("subjectName"));
+          item.setDate(new Date());
+          
+          restItems.add(item);
+        }
+        
+        webSocketMessenger.sendMessage("hops:workspace-signup", restItems, Arrays.asList(studentEntity, sessionController.getLoggedUserEntity()));
+
+        return Response.ok(restItems).build();
+      }
+      else {
+        throw new RuntimeException(String.format("Search provider couldn't find a unique workspace. %d results.", searchResult.getTotalHitCount()));
```

Comment:
Do not throw runtime exceptions in endpoints. Replace with something like `return Response.status(Status.INTERNAL_SERVER_ERROR).entity("yaddayaddayadda").build();`